### ISSUE TITLE
Pop back to first screen for a given tab on change

### ISSUE THREAD

**Issue by mwlabs42** [jb3rndt/PersistentBottomNavBarV2#33]
I apologize if this is redundant - I'm relatively new to Flutter.

I'm trying to get the following behavior to work when I change tabs: I'd like the state of each tab to be preserved (stateManagement : true), but the tab to be reverted to the first screen when the tabs are changed.

So what I'd like is (from README)

_Pop back to first screen in the navigation graph for a given tab:_

    Navigator.of(context).popUntil(ModalRoute.withName("/"));

However, I can't seem to get this work. I'm trying to override _onPressed_ in _PersistentBottomNavBarItem_. I'm not sure what I'm doing wrong - am I not getting the right _context_?

```DART
PersistentBottomNavBarItem(
  icon: Icon(icon),
  title: captionString,
  activeColorPrimary: Colors.blue,
  inactiveColorPrimary: Colors.grey,
  inactiveColorSecondary: Colors.purple,
  routeAndNavigatorSettings: RouteAndNavigatorSettings(
      onGenerateRoute: RouteGenerator.generateRoute),
  onPressed: (context) {
    if (_currentIndex != index && _childContext != null) {
      //this is not working
      if (Navigator.of(_childContext!).canPop())
        Navigator.of(_childContext!)
            .popUntil(ModalRoute.withName("/"));
    }
    setState(() {
      _currentIndex = index;
      _controller.index = _currentIndex;
      _childContext = context;
    });
    _populateChildWidget(entry.key);
  },
)
```

The error I keep getting (I've tried this in different ways) is

`'package:flutter/src/widgets/navigator.dart': Failed assertion: line 5140 pos 12: '_history.isNotEmpty': is not true.`

**Comment by jb3rndt** [comment]
Im not exactly sure if I understand your issue. As I understood, you want to have state management but also want to pop the whole navigator stack of the tab you are switching to. So you would effectively see the first page of each tab every time when switching to it. What effect would state management have in your case then?

Here is a way to achieve popping the navigator stack when switching to a route:
In your `PersistentTabView` set the `selectedTabScreenContext` attribute like this:
```dart
selectedTabScreenContext: (context) {
  currentContext = context;
  Future(() {
    if (Navigator.of(context).canPop())
      Navigator.of(context).popUntil(ModalRoute.withName("/"));
  });
}
```
Make sure that "/" is the initial route for every `PersistentBottomNavBarItem`.

Also, I suspect your example didnt work because the context is not right, as you already said. But I am not sure as I dont know the code around it.
And for clarification: The `onPressed` method on `PersistentBottomNavBarItem` prevents this tab to show the default page in your `screens` list. Instead, it executes the `onPressed` method, so you could show a dialog there.

While trying out the `selectedTabScreenContext` method myself, I found an issue with the method returning wrong contexts. I already fixed that, so thank you for finding that issue by mistake 😄.

**Comment by mwlabs42** [comment]
Thanks - yes, that's indeed the behavior I'm looking for. Setting state_mangement to false ends up calling initState for each widget, which is costly for my case.

I'm still not sure what the issue is, but I realized that both your code and mine works when I replace

`Navigator.of(context).popUntil(ModalRoute.withName("/"));`

with 

`
 Navigator.of(context).popUntil(ModalRoute.withName(
                      '/9f580fc5-c252-45d0-af25-9429992db112'))
`

Basically I tried to modify your code and found this from popAllScreens(). No idea what's going, but there seems to be a related stackoverflow question.


https://stackoverflow.com/questions/70356653/persistent-bottom-navigation-bar-does-not-work-with-named-routes

**Comment by jb3rndt** [comment]
`'/9f580fc5-c252-45d0-af25-9429992db112'` is used as a default for initialRoute of each tabs Navigator, if nothing other is specified.
What I found out about this is that if I set initialRoute (e.g. `'/'`) on a Tab, doing `Navigator.of(context).popUntil(ModalRoute.withName('/9f580fc5-c252-45d0-af25-9429992db112'))` pops so there is nothing left (white screen) on that tab. Setting no initialRoute and doing `Navigator.of(context).popUntil(ModalRoute.withName("/"));` results in the same.
So as this seems to be error prone, the best would be to use the initialRoute you specified in your code, I guess.
